### DISCUSSION TITLE
How can I use "printf"?

### DISCUSSION THREAD

**Discussion by rempas** [vlang/v#13692] [General]
Until string interpolation gets support for the native backend and I can use strings, how can I use `printf`?

The problem is in the declaration of the `printf` as V doesn't allow variadic functions that take arguments of different types. So the following won't work:

```
fn C.printf(format ...&char) int
C.printf(c"Age: %d", 10)
```

and I'll get the following error:

```
error: cannot use `int` as `&char` in argument 2 to `C.printf`
```

I have also tried to make do the following (which I know doesn't make sense but what can I do? I try everything...):

```
fn C.printf(format ...voidptr) int
C.printf(c"Age: %d", 10)
```
And I get:

```
native error: opcodes format: xx xx xx xx
```

Any ideas?

**EDIT:** The second decoration works when I use the C backend. So is this a bug to the native backend? Should I report it?

**Comment by Wajinn** [comment]
Happen to read this, and I'm a bit confused by it.  I thought string interpolation in V was already usable, so not understanding the need for C's printf as shown versus V's println.

```
a := 'hello'
b := 'hola'
println('$a is $b')
```
"V doesn't allow variadic functions that take arguments of different types."

Actually, **V does allow such**.  For this we would use Sum Types.  This also might be partially overlooked by those coming from OOP based languages where it doesn't have such or usage is very convoluted or more obscure.  Functional programming languages would usually make usage and clarification of Sum Types and Product Types more obvious.

```
type Mixer = int | string | bool   // put the different types or structs that you want

fn main() {
	println(mix_type("Hello", "John", 10, true))  
	println(generics_mix("Tom", "Derrick", "Harry"))
	println(generics_mix(25, 45, 72))
}

// function is accepting arguments of different types

fn mix_type(list ...Mixer) ([]string, []int, []bool) {
	mut nums := []int{}
	mut strs := []string{}
	mut bools := []bool{}
	
	for value in list {
		if value is int {
			nums << value
		}
		if value is string {
			strs << value
		}
		if value is bool	{
			bools << value
		}
	}
	return strs, nums, bools
}

// generic variadic example here for comparison; any type, but all parameters must be of that type

fn generics_mix<T>(args ...T) []T {
	for arg in args {
		print('Args: $arg \n')
	}
	return args
}
```

**Reply by rempas** [reply]
String interpolation doesn't work with native yet. For the other problem, sum types are great but I have to manually add the types I want to use so it won't do. And for generic variadic, it will also won't do because like I said, I need the arguments to be of a different type.

**Reply by Wajinn** [reply]
But if you don't use sum types or generics, in a strongly and statically typed language, then it would seem to be defeating the purpose.  Using `...any` or variant variables with variadic could introduce all kinds of run-time errors and issues in many situations.  Even in languages where they have this type of variant variable, despite being strongly and statically typed otherwise, the use of generics (to include with constraints) is usually recommended and prevalent.  In modern practice, usually variant variables are reserved for the more extreme and truly odd cases.

**Reply by rempas** [reply]
Yeah, probably the only case I needed variadic arguments of different types was for "printf" or "scanf".

**Comment by JalonSolov** [comment]
Short answer to the original question:  You can't.  At least not yet.

The `-native` code generation isn't mature enough, yet, to handle most pure V code, let alone C interop.  Doing that will require calling a linker at the very least, which means implementing a way to call a linker directly, etc.

**Reply by rempas** [reply]
Yeah, I know that "native" is not mature yet and it makes sense tbh. I was just asking in case there was something I didn't knew.
